File: encoder.js
const fs = require("fs");
const path = require("path");
const { Client } = require("ssh2");
const SFTPClient = require("ssh2-sftp-client");

const configuration = JSON.parse(fs.readFileSync(path.join(__dirname, "config.json"), "utf8"));
if (!configuration.ssh || !configuration.ssh.hostname || !configuration.ssh.port || !configuration.ssh.username || !configuration.ssh.password) {
    console.error("i1DT - SSH configuration is missing in config.json");
    process.exit(1);
}
const { hostname, port, username, password } = configuration.ssh;
    
const sshConfig = {
    host: hostname,
    port: port,
    username: username,
    password: password,
    algorithms: {
        kex: ["diffie-hellman-group-exchange-sha1", "diffie-hellman-group1-sha1"],
        serverHostKey: ["ssh-rsa", "ssh-dss"],
        cipher: ["aes256-cbc", "aes192-cbc", "aes128-cbc", "3des-cbc"],
    },
};

const conn = new Client();
let shellStream = null;
let sshConnected = false;

if (!fs.existsSync(path.join(__dirname, "temp"))) {
    fs.mkdirSync(path.join(__dirname, "temp"));
}

function connectSSH() {
    conn.on("ready", () => {
        console.log("i1DT - SSH connection established.");
        conn.shell((err, stream) => {
            if (err) throw err;
            shellStream = stream;
            sshConnected = true;
            stream.on("close", () => {
                console.log("i1DT - SSH stream closed");
                sshConnected = false;
                conn.end();
            })
            .on("data", (data) => {
                if (data.toString().includes("Password:")) {
              stream.write("i1\n");
                }
            });
            stream.write("su -l dgadmin\n");
        });
    })
    .connect(sshConfig);
}

function sendCommand(command) {
    if (!sshConnected) connectSSH();
    if (!shellStream) {
        console.log("i1DT - SSH session not ready yet.");
        return;
    }
    console.log(`i1DT - Sent command: ${command}`);
    shellStream.write(command + "\n");
}

async function getConfig() {
    const sftp = new SFTPClient();
    try {
        await sftp.connect(sshConfig);
        await sftp.get("/usr/home/dgadmin/config/current/config.py", path.join(__dirname, "config.py"));
        console.log("i1DT - Config downloaded.");
        sftp.end();

        const config = fs.readFileSync(path.join(__dirname, "config.py"), "utf8");
        const locationRegex = /wxdata\.setInterestList\('coopId','1',\[(.*?)\]\)/g;
        const locations = [];
        let match;
        while ((match = locationRegex.exec(config)) !== null) {
            const locationIds = match[1].split(",").map((id) => id.trim().replace(/['"]/g, ""));
            const filteredIds = locationIds.filter((id) => id !== "");
            locations.push(...filteredIds);
        }

        const locationRegexTecci = /wxdata\.setInterestList\('obsStation','1',\[(.*?)\]\)/g;
        const locationsTecci = [];
        let matchTecci;
        while ((matchTecci = locationRegexTecci.exec(config)) !== null) {
            const locationIds = matchTecci[1].split(",").map((id) => id.trim().replace(/['"]/g, ""));
            const filteredIds = locationIds.filter((id) => id !== "");
            const filteredIdsFinal = filteredIds.filter((id) => !(id.length === 4 && (id.startsWith("K") || id.startsWith("W"))));
            locationsTecci.push(...filteredIdsFinal);
        }
        const uniqueLocations = [...new Set(locations)];
        const uniqueLocationsTecci = [...new Set(locationsTecci)];
        const configData = { ssh: configuration.ssh, coop: { locations: uniqueLocations }, tecci: { locations: uniqueLocationsTecci } };
        fs.writeFileSync(path.join(__dirname, "config.json"), JSON.stringify(configData, null, 2));
        console.log("i1DT - Config parsed and config.json written.");
        return configData;
    } catch (err) {
        console.error("i1DT - Error fetching config:", err);
    }
}

async function getData(record, locations) {
    try {
        const res = await fetch(`https://wist.minnwx.com/api/i1/${record}/${locations.join(",")}?apiKey=5da46bbb68c49e48e05dc3362ea65adf`);
        const text = await res.text();
        fs.writeFileSync(path.join(__dirname, "temp", `${record}.py`), text);

        const sftp = new SFTPClient();
        await sftp.connect(sshConfig);
        await sftp.put(path.join(__dirname, "temp", `${record}.py`), `/home/dgadmin/${record}.py`);
        console.log(`i1DT - ${record}.py uploaded.`);

        if (!sshConnected) connectSSH();
        setTimeout(() => sendCommand("su -l dgadmin"), 1000);
        setTimeout(() => sendCommand("source ~/.bash_profile"), 1500);
        setTimeout(() => sendCommand(`runomni /twc/util/loadSCMTconfig.pyc /home/dgadmin/${record}.py`), 2000);
        sftp.end();
    } catch (err) {
        console.error(`i1DT - Error processing ${record}:`, err);
    }
}

async function startSchedules() {
    connectSSH();
    const config = await getConfig();
    if (!config || !config.coop || !config.coop.locations) {
        console.error("i1DT - Failed to load config.");
        return;
    }

    getData("cc", config.tecci.locations);
    getData("hourly", config.coop.locations);
    getData("daily", config.coop.locations);
    getData("daypart", config.coop.locations)

    setInterval(() => getData("cc", config.tecci.locations), 10 * 60 * 1000);

    setInterval(() => {
        getData("hourly", config.coop.locations);
        getData("daily", config.coop.locations);
        getData("daypart", config.coop.locations)
    }, 30 * 60 * 1000);

    console.log("i1DT - Data fetch schedule started.");
}

// Start process loop
startSchedules();

File: encoder.py
import paramiko
import os
import re
import json
import time
import threading
from datetime import datetime, timedelta

import cc
import hourly
import daily
import daypart

with open("config.json", "r") as f:
    ssh_config = json.load(f).get("ssh", {})

ssh_connected = False
ssh_client = None
shell = None


def ensure_temp_dir():
    if not os.path.exists("temp"):
        os.makedirs("temp")


def connect_ssh():
    global ssh_client, shell, ssh_connected
    ssh_client = paramiko.SSHClient()
    ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    ssh_client.connect(
        ssh_config["hostname"],
        port=ssh_config["port"],
        username=ssh_config["username"],
        password=ssh_config["password"],
        look_for_keys=False,
        allow_agent=False
    )
    ssh_connected = True
    shell = ssh_client.invoke_shell()
    print("i1DT - SSH connection established.")

    def handle_output():
        while True:
            output = shell.recv(1024).decode()
            if "Password:" in output:
                shell.send("i1\n")

    threading.Thread(target=handle_output, daemon=True).start()
    shell.send("su -l dgadmin\n")


def send_command(command):
    if not ssh_connected:
        connect_ssh()
    print(f"i1DT - Sent command: {command}")
    shell.send(command + "\n")

def sync_time():
    now = datetime.now()
    freebsd_timestamp = now.strftime("%m%d%H%M%Y.%S") # Generate current FreeBSD timestamp
    print("i1DT - Syncing Time, Timestamp is:" + freebsd_timestamp)
    send_command("date " + freebsd_timestamp) # Sync the time of the VM

def get_config():
    ensure_temp_dir()
    transport = paramiko.Transport((ssh_config["hostname"], ssh_config["port"]))
    transport.connect(username=ssh_config["username"], password=ssh_config["password"])
    sftp = paramiko.SFTPClient.from_transport(transport)

    local_path = "config.py"
    remote_path = "/usr/home/dgadmin/config/current/config.py"
    sftp.get(remote_path, local_path)
    print("i1DT - Config downloaded from i1.")
    sftp.close()
    transport.close()

    with open(local_path, "r") as f:
        config_content = f.read()

    coop_pattern = r"wxdata\.setInterestList\('coopId','1',\[(.*?)\]\)"
    coop_matches = re.findall(coop_pattern, config_content)
    locations = []
    for match in coop_matches:
        ids = [id.strip().replace("'", "").replace('"', '') for id in match.split(",") if id.strip()]
        ids = [id for id in ids if not id.startswith(("K", "W"))]
        locations.extend(ids)

    tecci_pattern = r"wxdata\.setInterestList\('obsStation','1',\[(.*?)\]\)"
    tecci_matches = re.findall(tecci_pattern, config_content)
    locations_tecci = []
    for match in tecci_matches:
        ids = [id.strip().replace("'", "").replace('"', '') for id in match.split(",") if id.strip()]
        ids = [id for id in ids if not id.startswith(("K", "W"))]
        locations_tecci.extend(ids)

    config_data = {
        "ssh": ssh_config,
        "coop": {"locations": list(set(locations))},
        "tecci": {"locations": list(set(locations_tecci))}
    }

    with open("config.json", "w") as f:
        json.dump(config_data, f, indent=2)

    print("i1DT - Config parsed and config.json written.")
    return config_data


def upload_and_run_temp_files():
    ensure_temp_dir()
    transport = paramiko.Transport((ssh_config["hostname"], ssh_config["port"]))
    transport.connect(username=ssh_config["username"], password=ssh_config["password"])
    sftp = paramiko.SFTPClient.from_transport(transport)

    for file_name in os.listdir("temp"):
        local_path = os.path.join("temp", file_name)
        remote_path = f"/home/dgadmin/{file_name}"
        sftp.put(local_path, remote_path)
        print(f"i1DT - Uploaded {file_name}")

        if not ssh_connected:
            connect_ssh()

        time.sleep(0.5)
        send_command(f"runomni /twc/util/loadSCMTconfig.pyc {remote_path}")

    sftp.close()
    transport.close()


def start_schedules():
    config = get_config()
    if not config or "coop" not in config or "locations" not in config["coop"]:
        print("i1DT - Failed to load i1 config.")
        return

    def run_cc():
        while True:
            ensure_temp_dir()
            cc.main()
            upload_and_run_temp_files()
            time.sleep(600)

    def run_hourly_daily_daypart():
        while True:
            ensure_temp_dir()
            hourly.main()
            daily.main()
            daypart.main()
            upload_and_run_temp_files()
            time.sleep(1800)

    threading.Thread(target=run_cc, daemon=True).start()
    threading.Thread(target=run_hourly_daily_daypart, daemon=True).start()

    print("i1DT - Data generation & upload schedules started.")


if __name__ == "__main__":
    sync_time()
    start_schedules()
    while True:
        time.sleep(1)
